akka-sample-persistence-dc-scala/src/main/scala/sample/persistence/multidc/ThumbsUpCounter.scala

```diff
@@ -1,77 +1,66 @@
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.{ActorRef, Behavior}
+import akka.cluster.sharding.typed.{ReplicatedEntityProvider, ShardingEnvelope}
+import akka.persistence.cassandra.query.scaladsl.CassandraReadJournal
+import akka.persistence.typed.scaladsl.{Effect, EventSourcedBehavior, ReplicatedEventSourcing}
+import akka.persistence.typed.{ReplicaId, ReplicationId}
 
 object ThumbsUpCounter {
-  sealed trait Command {
+
+  // sent over sharding
+  sealed trait Command extends CborSerializer {
     def resourceId: String
   }
 
-  final case class GiveThumbsUp(resourceId: String, userId: String) extends Command
+  final case class GiveThumbsUp(resourceId: String, userId: String, replyTo: ActorRef[Long]) extends Command
 
-  final case class GetCount(resourceId: String) extends Command
+  final case class GetCount(resourceId: String, replyTo: ActorRef[Long]) extends Command
 
-  final case class GetUsers(resourceId: String) extends Command
+  final case class GetUsers(resourceId: String, replyTo: ActorRef[State]) extends Command
 
-  sealed trait Event
+  // saved to DB
+  sealed trait Event extends CborSerializer
 
   final case class GaveThumbsUp(userId: String) extends Event
 
-  final case class State(users: Set[String]) {
+  // saved to DB
+  final case class State(users: Set[String]) extends CborSerializer {
     def add(userId: String): State = copy(users + userId)
   }
 
-  def shardingProps(settings: PersistenceMultiDcSettings): Props =
-    ReplicatedEntity.clusterShardingProps(ShardingTypeName, () => new ThumbsUpCounter, settings)
-
-  val ShardingTypeName = "counter"
-
-  val extractEntityId: ShardRegion.ExtractEntityId = {
-    case cmd: Command => (cmd.resourceId, cmd)
-    case evt: SpeculativeReplicatedEvent => (evt.entityId, evt)
-  }
-
-  val MaxShards = 100
-  def shardId(entityId: String): String = (math.abs(entityId.hashCode) % MaxShards).toString
-  val extractShardId: ShardRegion.ExtractShardId = {
-    case cmd: Command   => shardId(cmd.resourceId)
-    case evt: SpeculativeReplicatedEvent => shardId(evt.entityId)
-    case StartEntity(entityId)           => shardId(entityId)
-  }
-}
-
-class ThumbsUpCounter
-  extends ReplicatedEntity[ThumbsUpCounter.Command, ThumbsUpCounter.Event, ThumbsUpCounter.State] {
-  import ThumbsUpCounter._
-
-  override def initialState: State = State(Set.empty)
+  // In this sample one replica per DC
+  val Replicas = Set(ReplicaId("eu-west"), ReplicaId("eu-central"))
+
+  val Provider: ReplicatedEntityProvider[Command] = ReplicatedEntityProvider.perDc("counter", Replicas) { replicationId => ThumbsUpCounter(replicationId) }
+
+
+  // we use a shared journal as cassandra typically spans DCs rather than a DB per replica
+  def apply(replicationId: ReplicationId): Behavior[Command] =
+    Behaviors.setup { ctx =>
+      ReplicatedEventSourcing.withSharedJournal(replicationId, Replicas, CassandraReadJournal.Identifier) { replicationContext =>
```

Comment:
What about this `withSharedJournal`, and there is one alternative. Could we simplify/collapse those?

The word "shared" here hasn't fully clicked for me. I think it's the journal configuration that is shared, right? When is that desired and when is it not shared?

Also easy to misread it as related to sharding.